src/txn/txn_ckpt.c

```diff
@@ -1139,9 +1140,20 @@ __drop(WT_CKPT *ckptbase, const char *name, size_t len)
      * it's one we want to drop.
      */
     if (strncmp(WT_CHECKPOINT, name, len) == 0) {
```

Comment:
It's feature creep, but should we also skip this entire step if a hot backup is active? We aren't ever going to drop those checkpoints in that case.

Reply:
We could, but we're just skipping some number of short string comparisons, do you think we could detect the performance change?

Reply:
String comparisons and manipulations are expensive. If we can avoid a string comparison on every checkpoint, then I'm a fan.

Reply:
I think it's worthwhile, but feel free to push that out of here if you'd like further discussion. It is definitely feature creep.

Reply:
@agorrod, @sueloverso:

I must be missing something because I don't see it.

RIght now there are two steps.

In step #1 we walk the list of checkpoints for a tree and flag checkpoints we're going to drop (that's the three `__drop` functions), and that's where the string comparisons happen. There are multiple functions because there are different reasons to drop checkpoints, for example, doing a named checkpoint where the named checkpoint already exists.

Once we flag the checkpoints to drop in step #1,  we do a second pass over the list of checkpoints. In this step #2, if the hot-backup cursor is open, we either fail the checkpoint (when the checkpoint can't proceed), or clear the checkpoint's delete flag, if we can proceed but we can't delete the checkpoint.

We could avoid both of these steps, but what we're saving is 4 (`WT_MAX_CHECKPOINT_DROP`) string comparisons in step #1, and a full walk of the checkpoint array in step #2. Note, in that full walk we're not doing string comparisons, we're testing and clearing a flag in the checkpoint structure.

Admittedly, there could be a lot of checkpoint structures if the hot backup cursor has been open  for a long time.

It seems to me to do this we'll have to split apart the drop functions (some drops have to happen anyway, it's just the "standard" checkpoints where the drops no longer have to happen), that is, there's a set of drop checks we've already done, and a set of other drop checks that still have to happen, and that just feels more fragile/complicated to me, and I don't see the performance win.

Can one of you please help me understand what I'm missing?

Reply:
I was just saying that generally string stuff is expensive. If it's easy to avoid great. What you described (I didn't look at all the paths in this section of code) sounds too complicated to be much benefit right now.